runners/reference/java/src/main/java/org/apache/beam/runners/reference/PortableRunner.java

```diff
@@ -0,0 +1,245 @@
+
+    JobServiceBlockingStub jobService = JobServiceGrpc.newBlockingStub(jobServiceChannel);
+    try (CloseableResource<JobServiceBlockingStub> wrappedJobService =
+        CloseableResource.of(jobService, (unused) -> jobServiceChannel.shutdown())) {
+
+      PrepareJobResponse prepareJobResponse = jobService.prepare(prepareJobRequest);
+      LOG.info("PrepareJobResponse: {}", prepareJobResponse);
+
+      ApiServiceDescriptor artifactStagingEndpoint =
+          prepareJobResponse.getArtifactStagingEndpoint();
+
+      String stagingToken = null;
+      try (CloseableResource<ManagedChannel> artifactChannel =
+          CloseableResource.of(
+              channelFactory.forDescriptor(artifactStagingEndpoint), ManagedChannel::shutdown)) {
+        ArtifactServiceStager stager = ArtifactServiceStager.overChannel(artifactChannel.get());
+        LOG.debug("Actual files staged: {}", filesToStage);
+        stagingToken = stager.stage(filesToStage);
+      } catch (CloseableResource.CloseException e) {
+        LOG.warn("Error closing artifact staging channel", e);
+        // CloseExceptions should only be thrown while closing the channel.
+        checkState(stagingToken != null);
+      } catch (Exception e) {
+        throw new RuntimeException("Error staging files.", e);
+      }
+
+      RunJobRequest runJobRequest =
+          RunJobRequest.newBuilder()
+              .setPreparationId(prepareJobResponse.getPreparationId())
+              .setStagingToken(stagingToken)
+              .build();
+
+      RunJobResponse runJobResponse = jobService.run(runJobRequest);
+
+      LOG.info("RunJobResponse: {}", runJobResponse);
+      ByteString jobId = runJobResponse.getJobIdBytes();
+
+      return new JobServicePipelineResult(jobId, wrappedJobService.transfer());
+    } catch (CloseException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public String toString() {
+    return "PortableRunner#" + hashCode();
+  }
+
+  private static File zipDirectory(File directory) throws IOException {
+    File zipFile = File.createTempFile(directory.getName(), ".zip");
+    try (FileOutputStream fos = new FileOutputStream(zipFile)) {
+      ZipFiles.zipDirectory(directory, fos);
+    }
+    return zipFile;
+  }
+
+  private static StagedFile createStagingFile(File file) {
+    // https://issues.apache.org/jira/browse/BEAM-4109
+    // HACK: Encode the path name ourselves because the local artifact staging service currently
+    // assumes artifact names correspond to a flat directory. Artifact staging services should
+    // generally accept arbitrary artifact names.
+    // NOTE: Base64 url encoding does not work here because the stage artifact names tend to be long
+    // and exceed file length limits on the artifact stager.
+    String encodedPath = escapePath(file.getPath());
+    return StagedFile.of(file, encodedPath);
+  }
+
+  /** Create a filename-friendly artifact name for the given path. */
```

Comment:
<!--thread_id:cc_182613993_t; commit:a16a978c74d440d81ca78b3c44be72eae312ef74; resolved:1-->
<!--section:context-quote-->
> **tgroh** wrote:
> It really seems like there's gotta be either a library way to do this, or some clever workaround?

<!--section:body-->
This _is_ the clever workaround.

I actually tried base64 encoding, but it turns out that very long path names end up on the classpath and base64 encoding causes path names to exceed size limits on the systems I tried it on. See https://github.com/apache/beam/commit/13ce9cf872e2856fbddff6822b5017a53bbd90f6.

The proper fix is to support arbitrary staging names server-side by maintaining an artifact name mapping independent of filesystem rules and limits. I created BEAM-4109 to track this; it should really be its own change.